### PR TITLE
Pythia external access

### PR DESCRIPTION
**Pull request by tmat**
Implements #35495, but does not restrict the IVTs yet.

Fixes https://dev.azure.com/devdiv/DevDiv/_workitems/edit/1050791

### CODE REVIEW

**Comment by tmat** [comment]
@sharwell PTAL

File: src/Workspaces/Core/Portable/ExternalAccess/Pythia/Api/PythiaAsyncLazyWrapper.cs [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+#nullable enable
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.Pythia.Api
+{
+    internal struct PythiaAsyncLazyWrapper<T>
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/ExternalAccess/Pythia/Api/PythiaAsyncLazyWrapper.cs]: ❔ Why is this needed? Pythia would already have access to the `AsyncLazy<T>` defined in vs-threading, which is preferable for use in Visual Studio.

File: src/Workspaces/Core/Portable/ExternalAccess/Pythia/Api/PythiaObjectPool.cs [hunk 2]
```diff
@@ -0,0 +1,27 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using Microsoft.CodeAnalysis.PooledObjects;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.Pythia.Api
+{
+    internal static class PythiaObjectPool
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/ExternalAccess/Pythia/Api/PythiaObjectPool.cs]: ⚠️  External code should not be using our object pools. Is this absolutely required, or is it sufficient to copy the sources for the pool into the external library for full encapsulation?

File: src/Workspaces/Core/Portable/ExternalAccess/Pythia/Api/PythiaSyntaxFactsServiceWrapper.cs [hunk 3]
```diff
@@ -0,0 +1,20 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.Pythia.Api
+{
+    internal struct PythiaSyntaxFactsServiceWrapper
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/ExternalAccess/Pythia/Api/PythiaSyntaxFactsServiceWrapper.cs]: 📝 All these new structs can be `readonly`

File: src/Features/Core/Portable/ExternalAccess/Pythia/Api/PythiaSymbolMatchPriority.cs [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+#nullable enable
+
+using Microsoft.CodeAnalysis.Completion.Providers;
+
+namespace Microsoft.CodeAnalysis.ExternalAccess.Pythia.Api
+{
+    internal static class PythiaSymbolMatchPriority
+    {
+        internal static int Keyword = SymbolMatchPriority.Keyword;
```
Comment (sharwell) [inline src/Features/Core/Portable/ExternalAccess/Pythia/Api/PythiaSymbolMatchPriority.cs]: 💡 Make these read-only and/or constant

File: src/Workspaces/Core/Portable/ExternalAccess/Pythia/Api/PythiaObjectPool.cs (hunk 2 above)
Comment (tmat) [reply src/Workspaces/Core/Portable/ExternalAccess/Pythia/Api/PythiaObjectPool.cs]: So, ideally Pythia would use Microsoft.CodeAnalysis.PooledObject source package. The problem is that since it has IVT to Roslyn the compiler sees both source implementation and implementation from metadata and reports warning. We can address that in two ways 1) make a copy and rename the namespace, 2) use extern alias for Roslyn references.

File: src/Features/Core/Portable/Completion/Providers/SymbolMatchPriority.cs [hunk 5]
```diff
@@ -1,6 +1,10 @@
...
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+#nullable enable
+
+namespace Microsoft.CodeAnalysis.Completion.Providers
 {
-    internal class SymbolMatchPriority
+    internal static class SymbolMatchPriority
     {
         internal static int Keyword = 100;
```
Comment (sharwell) [inline src/Features/Core/Portable/Completion/Providers/SymbolMatchPriority.cs]: 📝 All of these should have been `readonly`

File: src/Workspaces/Core/Portable/ExternalAccess/Pythia/Api/PythiaSyntaxFactsServiceWrapper.cs (hunk 3 above)
Comment (tmat) [reply src/Workspaces/Core/Portable/ExternalAccess/Pythia/Api/PythiaSyntaxFactsServiceWrapper.cs]: Yup.

File: src/Workspaces/Core/Portable/ExternalAccess/Pythia/Api/PythiaAsyncLazyWrapper.cs (hunk 1 above)
Comment (tmat) [reply src/Workspaces/Core/Portable/ExternalAccess/Pythia/Api/PythiaAsyncLazyWrapper.cs]: I guess I can switch them to VS AsyncLazy. The only issue is that it takes JoinableTaskFactory optional parameter and if I don't pass it I get `VSTHRD012: Provide JoinableTaskFactory where allowed`.  I guess I can ignore that since there is the code is not doing anything with UI.

File: src/Features/Core/Portable/ExternalAccess/Pythia/Api/PythiaSymbolMatchPriority.cs (hunk 4 above)
Comment (tmat) [reply src/Features/Core/Portable/ExternalAccess/Pythia/Api/PythiaSymbolMatchPriority.cs]: Making it read-only. Constant would be embedded to the target assembly, not sure if we want that.

File: src/Workspaces/Core/Portable/ExternalAccess/Pythia/Api/PythiaObjectPool.cs (hunk 2 above)
Comment (tmat) [reply src/Workspaces/Core/Portable/ExternalAccess/Pythia/Api/PythiaObjectPool.cs]: I'll keep it this way for now as it preserves the current behavior. We can remove later.
